### PR TITLE
RROA Polref interface: Add global options to settings tab

### LINKED ISSUE #16983: Add global options to `Settings` tab
This issue is to:
- Gather requirements from scientists about the global options that should be shown in the `Settings` tab. Edit: I talked to them about this and a good starting point is the presentation sent by Max (but see also list below):

settingstab
- Use a `QGridLayout` to put things in the correct place
- Move the existing `HintingLineEdit`s to the end.
- Forget about `Transmission Runs` for the moment.
- For each new feature, add unit tests for the `Settings` presenter
- For the moment, forget about the "Get defaults" buttons.

### PR DESCRIPTION
**Pull request by PranavBahuguna**
As described in the issue, this PR adds several settings to the settings tab. In the original interface, these settings were all specified together for each algorithm as keyword arguments. Now, each setting is given their own edit space and are split between 'Experiment' and 'Instrument' sections. In addition, a 
'Get Defaults' button has been added for each section that fills the settings with default values. For experiment settings, default values are pulled from the RROA algorithm whereas for instrument settings, they are pulled from the current instrument being used in the run.

The list of settings added are as follows:

Experiment settings:
- AnalysisMode
- DirectBeam
- Transmission run(s)
- StartOverlap
- EndOverlap
- PolarisationCorrections
- CRho
- CAlpha
- CAp
- CPp
- dQ/Q
- Scale
- Stitch1DMany

Instrument settings:
- IntegratedMonitors
- MonitorIntegralMin
- MonitorIntegralMax
- MonitorBackgroundMin
- MonitorBackgroundMax
- LambdaMin
- LambdaMax
- I0MonitorIndex
- ProcessingInstructions

Original interface (Global options) has been left in for now.

**To test:**

Ensure all unit tests pass.

Ensure that altering the settings from both experiment and instrument settings affect the algorithm runs correctly.

Ensure the 'Get Defaults' buttons for each section works as described. For `Instrument Settings' section, ensure it obtains the correct default values for each available instrument (set in the runs tab).

Fixes #16983.

Release notes have been updated
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [x] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [x] Do changes function as described? Add comments below that describe the tests performed?
- [x] How do the changes handle unexpected situations, e.g. bad input?
- [x] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: MantidQt/CustomInterfaces/inc/MantidQtCustomInterfaces/Reflectometry/IReflSettingsTabView.h [hunk 1]
```diff
@@ -87,8 +88,9 @@ class DLLExport IReflSettingsTabView {
   virtual void requestExpDefaults() const = 0;
   virtual void requestInstDefaults() const = 0;
   /// Set default values for settings
-  virtual void setExpDefaults(std::vector<std::string> defaults) const = 0;
-  virtual void setInstDefaults(std::vector<double> defaults) const = 0;
+  virtual void setExpDefaults(
+    const std::vector<std::string> defaults) const = 0;
+  virtual void setInstDefaults(const std::vector<double> defaults) const = 0;
```
Comment (raquelalvarezbanos) [inline MantidQt/CustomInterfaces/inc/MantidQtCustomInterfaces/Reflectometry/IReflSettingsTabView.h]: `const std::vector<double> &`?

File: MantidQt/CustomInterfaces/inc/MantidQtCustomInterfaces/Reflectometry/IReflSettingsTabPresenter.h [hunk 2]
```diff
@@ -46,6 +46,11 @@ class IReflSettingsTabPresenter {
   virtual std::string getReductionOptions() const = 0;
   /// Post-processing
   virtual std::string getStitchOptions() const = 0;
+
+  enum Flag { ExpDefaultsFlag, InstDefaultsFlag };
+
+  // Tell the presenter something happened
+  virtual void notify(IReflSettingsTabPresenter::Flag flag) = 0;
 };
```
Comment (raquelalvarezbanos) [inline MantidQt/CustomInterfaces/inc/MantidQtCustomInterfaces/Reflectometry/IReflSettingsTabPresenter.h]: Good :+1:

File: MantidQt/CustomInterfaces/inc/MantidQtCustomInterfaces/Reflectometry/IReflSettingsTabView.h [hunk 3]
```diff
@@ -62,6 +63,34 @@ class DLLExport IReflSettingsTabView {
...
+  virtual std::string getMonitorIntegralMax() const = 0;
+  virtual std::string getMonitorBackgroundMin() const = 0;
+  virtual std::string getMonitorBackgroundMax() const = 0;
+  virtual std::string getLambdaMin() const = 0;
+  virtual std::string getLambdaMax() const = 0;
+  virtual std::string getI0MonitorIndex() const = 0;
+
+  /// Request presenter to obtain default values for settings
+  virtual void requestExpDefaults() const = 0;
+  virtual void requestInstDefaults() const = 0;
```
Comment (raquelalvarezbanos) [inline MantidQt/CustomInterfaces/inc/MantidQtCustomInterfaces/Reflectometry/IReflSettingsTabView.h]: I don't think we need these methods to be defined in the abstract base class. In our abstract view we only need to define methods that will be used by the concrete presenter. It is enought to define them in the concrete view, as you've done below. When removed, they should be removed from the `MockSettingsTabView` too.

File: MantidQt/CustomInterfaces/inc/MantidQtCustomInterfaces/Reflectometry/IReflSettingsTabView.h [hunk 4]
```diff
@@ -62,6 +63,34 @@ class DLLExport IReflSettingsTabView {
...
+  /// Set default values for settings
+  virtual void
+  setExpDefaults(const std::vector<std::string> defaults) const = 0;
+  virtual void setInstDefaults(const std::vector<double> defaults) const = 0;
```
Comment (raquelalvarezbanos) [inline MantidQt/CustomInterfaces/inc/MantidQtCustomInterfaces/Reflectometry/IReflSettingsTabView.h]: These two are OK, because they are used by the presenter :+1:

File: MantidQt/CustomInterfaces/inc/MantidQtCustomInterfaces/Reflectometry/QtReflSettingsTabView.h [hunk 5]
```diff
@@ -52,6 +52,43 @@ class QtReflSettingsTabView : public QWidget, public IReflSettingsTabView {
...
+  std::string getMonitorBackgroundMax() const override;
+  /// Return wavelength min
+  std::string getLambdaMin() const override;
+  /// Return wavelength max
+  std::string getLambdaMax() const override;
+  /// Return I0MonitorIndex
+  std::string getI0MonitorIndex() const override;
+  /// Set default values for experiment and instrument settings
+  void setExpDefaults(const std::vector<std::string> defaults) const override;
+  void setInstDefaults(const std::vector<double> defaults) const override;
```
Comment (raquelalvarezbanos) [inline MantidQt/CustomInterfaces/inc/MantidQtCustomInterfaces/Reflectometry/QtReflSettingsTabView.h]: `const std::vector<double> &`?

File: MantidQt/CustomInterfaces/src/Reflectometry/ReflMainWindowPresenter.cpp [hunk 6]
```diff
@@ -139,6 +139,11 @@ ReflMainWindowPresenter::runPythonAlgorithm(const std::string &pythonCode) {
   return m_view->runPythonAlgorithm(pythonCode);
 }
 
+std::string ReflMainWindowPresenter::getInstrumentName() const {
+
+  return m_runsPresenter->getCurrentInstrumentName();
+}
+
```
Comment (raquelalvarezbanos) [inline MantidQt/CustomInterfaces/src/Reflectometry/ReflMainWindowPresenter.cpp]: Good. Our main presenter forwards the call to the `RunsPresenter`.

File: MantidQt/CustomInterfaces/src/Reflectometry/ReflRunsTabPresenter.cpp [hunk 7]
```diff
@@ -202,7 +202,7 @@ void ReflRunsTabPresenter::populateSearch(IAlgorithm_sptr searchAlg) {
   if (searchAlg->isExecuted()) {
     ITableWorkspace_sptr results = searchAlg->getProperty("OutputWorkspace");
     m_searchModel = ReflSearchModel_sptr(new ReflSearchModel(
-        *getTransferStrategy(), results, m_view->getSearchInstrument()));
+        *getTransferStrategy(), results, getCurrentInstrumentName()));
```
Comment (raquelalvarezbanos) [inline MantidQt/CustomInterfaces/src/Reflectometry/ReflRunsTabPresenter.cpp]: :+1:

File: MantidQt/CustomInterfaces/test/ReflSettingsTabPresenterTest.h [hunk 8]
```diff
@@ -65,6 +191,33 @@ class ReflSettingsTabPresenterTest : public CxxTest::TestSuite {
...
   }
+
+  void testExperimentDefaults() {
+    MockSettingsTabView mockView;
+    ReflSettingsTabPresenter presenter(&mockView);
+
+    std::vector<std::string> defaults = {"PointDetectorAnalysis", "None", "1"};
+
+    EXPECT_CALL(mockView, setExpDefaults(defaults)).Times(1);
+    presenter.notify(IReflSettingsTabPresenter::ExpDefaultsFlag);
```
Comment (raquelalvarezbanos) [inline MantidQt/CustomInterfaces/test/ReflSettingsTabPresenterTest.h]: I suggest adding here the following line:

```
TS_ASSERT(Mock::VerifyAndClearExpectations(&mockView));
```

that way we can verify the expectations on our `mockView` at this point. Please add this to all tests in `ReflSettingsTabPresenterTest` that are missing it.

**Comment by raquelalvarezbanos** [comment]
Code changes look very good. Good job @PranavBahuguna. 

I also tested functionality and everything works as I would expect. I am marking this PR with `Science Validation` as we should have a meeting with scientists to demo the new functionalily so they can give us their feedback. They may want to rename/add options.

**Comment by raquelalvarezbanos** [comment]
I noticed on Ubuntu, that the "Get Defaults" button does not display correctly. Could we try to fix it?

image 2

File: MantidQt/CustomInterfaces/test/ReflSettingsTabPresenterTest.h [hunk 9]
```diff
@@ -217,6 +217,7 @@ class ReflSettingsTabPresenterTest : public CxxTest::TestSuite {
         .WillOnce(Return("INTER"));
     EXPECT_CALL(mockView, setInstDefaults(defaults)).Times(1);
     presenter.notify(IReflSettingsTabPresenter::InstDefaultsFlag);
+    TS_ASSERT(Mock::VerifyAndClearExpectations(&mockView));
```
Comment (raquelalvarezbanos) [inline MantidQt/CustomInterfaces/test/ReflSettingsTabPresenterTest.h]: Thanks for adding this. I think the test `testExperimentDefaults` is also missing this line.

**Comment by raquelalvarezbanos** [comment]
Some modifications requested by scientists:

1. `Transmission run(s)` should be shown as in the original design document, but `TransmissionLambdaMin` and `TransmissionLambdaMax` shouldn't (we have to use `LambdaMin` and `LambdaMax` instead, as we are currently doing). In addition, we want to show `StartOverlap` and `EndOverlap`, which refer to transmission runs.
2. `MomentumTransferLimits`: they don't need the minimum and maximum Q values, only the step. This widget should be renamed to `dQ/Q` for consistency with the `Runs` tab.
3. `DetectorLimits` should be moved to `InstrumentSettings` and renamed to `ProcessingInstructions` to avoid confusion.
4. We need to add the property `NormalizeByIntegratedMonitors` (it will be a checkbox) to `InstrumentSettings`.
5. They don't need the section `Global Options` but they would like to keep the text box for `Stitch1DMany` which should be moved to `Experiment Settings`.
6. `PolarizationCorrections` should be greyed out when instrument is `INTER` or `SURF`. 

In summary, we should end up with something like this:

changes

@PranavBahuguna Item 6 is going to be a bit more complicated, so let me know when you are done with 5.

**Comment by raquelalvarezbanos** [comment]
@PranavBahuguna I've added two commits to implement functionality to load transmission runs and use them to reduce the data. When testing this, in addition to setting the corresponding expectation on the view as we've done with the rest of properties, we should also test that the transmission workspaces exist in the ADS.

File: MantidQt/CustomInterfaces/test/ReflSettingsTabPresenterTest.h [hunk 10]
```diff
@@ -144,6 +144,9 @@ class ReflSettingsTabPresenterTest : public CxxTest::TestSuite {
     EXPECT_CALL(mockView, getProcessingInstructions())
         .Times(Exactly(1))
         .WillOnce(Return("\"3,4\""));
+    EXPECT_CALL(mockView, getTransmissionRuns())
+        .Times(Exactly(1))
+        .WillOnce(Return("INTER00013463.nxs,INTER00013464.nxs"));
```
Comment (raquelalvarezbanos) [inline MantidQt/CustomInterfaces/test/ReflSettingsTabPresenterTest.h]: I suggest just using "INTER00013463, INTER00013464" here, or, even better, just "13463, 13464". The reason is that this way the test will reflect what will happen in reality, as users will only enter run numbers, not the full name and extension of the transmission workspace.

File: MantidQt/CustomInterfaces/test/ReflSettingsTabPresenterTest.h [hunk 11]
```diff
@@ -166,6 +169,8 @@ class ReflSettingsTabPresenterTest : public CxxTest::TestSuite {
     TS_ASSERT_EQUALS(optionsVec[15], "ScaleFactor=2");
     TS_ASSERT_EQUALS(optionsVec[16], "MomentumTransferStep=-0.02");
     TS_ASSERT_EQUALS(optionsVec[17], "ProcessingInstructions=\"3,4\"");
+    TS_ASSERT_EQUALS(optionsVec[18], "FirstTransmissionRun=TRANS_INTER00013463.nxs");
+    TS_ASSERT_EQUALS(optionsVec[19], "SecondTransmissionRun=TRANS_INTER00013464.nxs");
```
Comment (raquelalvarezbanos) [inline MantidQt/CustomInterfaces/test/ReflSettingsTabPresenterTest.h]: In addition to this, we should test that workspaces were created in the ADS, something like:

```
TS_ASSERT(AnalysisDataService::Instance().doesExist("TRANS_13464");
TS_ASSERT(AnalysisDataService::Instance().doesExist("TRANS_13464");
```

We should also clear the ADS by adding:

```
AnalysisDataService::Instance().clear();
```

File: MantidQt/CustomInterfaces/inc/MantidQtCustomInterfaces/Reflectometry/ReflSettingsTabPresenter.h [hunk 12]
```diff
@@ -72,6 +73,8 @@ class MANTIDQT_CUSTOMINTERFACES_DLL ReflSettingsTabPresenter
   IReflSettingsTabView *m_view;
   /// The main presenter
   IReflMainWindowPresenter *m_mainPresenter;
+  /// Name of the current instrument in use
+  std::string currentInstrumentName;
```
Comment (raquelalvarezbanos) [inline MantidQt/CustomInterfaces/inc/MantidQtCustomInterfaces/Reflectometry/ReflSettingsTabPresenter.h]: If this is a member variable it should be renamed to `m_currentInstrumentName`. See [here](http://www.mantidproject.org/C%2B%2B_Coding_Standards#Naming_Conventions) for more naming conventions in the project.

File: MantidQt/CustomInterfaces/src/Reflectometry/ReflMainWindowPresenter.cpp [hunk 13]
```diff
@@ -20,9 +20,6 @@ ReflMainWindowPresenter::ReflMainWindowPresenter(
   // Tell the tab presenters that this is going to be the main presenter
   m_runsPresenter->acceptMainPresenter(this);
   m_settingsPresenter->acceptMainPresenter(this);
-
-  // Trigger the setting of the current instrument name in settings tab
-  m_runsPresenter->notify(IReflRunsTabPresenter::InstrumentChangedFlag);
```
Comment (raquelalvarezbanos) [inline MantidQt/CustomInterfaces/src/Reflectometry/ReflMainWindowPresenter.cpp]: Why was this removed? My understanding is that this notification was triggering the `InstrumentChangedFlag` that would cause `PolarizationCorrections` to be enable/disable depending on the selected instrument.

Without this, if my default instrument is for example `INTER`, when I open the interface, even if `INTER` is the selected instrument in the `Runs` tab, in the `Settings` tab `PolarizationCorrections` is not greyed out. I think what you had here was correct.

File: MantidQt/CustomInterfaces/src/Reflectometry/ReflSettingsTabPresenter.cpp [hunk 14]
```diff
@@ -16,7 +16,7 @@ using namespace Mantid::Geometry;
 * @param view :: The view we are handling
 */
 ReflSettingsTabPresenter::ReflSettingsTabPresenter(IReflSettingsTabView *view)
-    : m_view(view), m_mainPresenter() {
+    : m_view(view), m_mainPresenter(), currentInstrumentName("INTER") {
```
Comment (raquelalvarezbanos) [inline MantidQt/CustomInterfaces/src/Reflectometry/ReflSettingsTabPresenter.cpp]: As discussed earlier, what will happen if my default instrument is for example `POLREF`? Will the `ReflSettingsTabPresenter` think that it is `INTER` and get defaults for `INTER` instead of `POLREF`?

**Comment by raquelalvarezbanos** [comment]
Looks very good to me. New options work as described.

Good job @PranavBahuguna

**Comment by raquelalvarezbanos** [comment]
:shipit:
